torch/lowrank.py

```diff
@@ -0,0 +1,249 @@
+"""Implement various linear algebra algorithms for low rank matrices.
+"""
+
+__all__ = ['svd', 'pca']
+
+import torch
+from . import _linalg_utils as _utils 
+
+
+def get_approximate_basis(A, q, niter=2, M=None):
+    """Return tensor :math:`Q` with q orthonormal columns such that
+    :math:`Q Q^H A` approximates :math:`A`. If :math:`M` is specified,
+    then `Q` is such that :math:`Q Q^H (A - M)` approximates :math:`A
+    - M`.
+
+    .. note:: The implementation is based on the Algorithm 4.4 from
+              Halko et al, 2009.
+
+    .. note:: For an adequate approximation of a k-rank matrix A,
+              where k is not known in advance but could be estimated,
+              the number of Q columns q can be choosen according to
+              the following criteria: in general, :math:`k <= q <=
+              min(2*k, m, n)`. For large low-rank matrices, take
+              :math:`q = k + 5..10`.  If k is relatively small
+              compared to :math:`min(m, n)`, choosing :math:`q = k +
+              0..2` may be sufficient.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int): the dimension of subspace spanned by Q columns.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer. In most cases, the default
+                               value 2 is more than enough.
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    dtype = _utils.get_floating_dtype(A)
+    matmul = _utils.get_matmul(A)
+
+    R = _utils.uniform(low=-1.0, high=1.0, size=(n, q),
+                       dtype=dtype, device=A.device)
```

Comment:
This matrix has to be Gaussian, not uniform. The upperbounds on errors work only because of concentration properties of (sub)Gaussian distributions. Uniforms does not concentrate.

Reply:
Changed to `torch.randn`.
Re using uniform: the algorithm appears to be robust to the distribution choice.